### DISCUSSION TITLE
How to use TimeStamps default class correctly

### DISCUSSION THREAD

**Discussion by wong2** [typegoose/typegoose#512] [Q&A] [answered]
```ts
class User extends defaultClasses.TimeStamps {
  @prop({ required: true })
  public name: string
}
```

The timestamps `createdAt` and `updatedAt` should be automatically created. However, they are marked optional in types, so I have to use `user.createdAt!.toString()` to use them in code, is there better ways?

**Comment by smolinari** [comment]
@wong2 

I don't use the TimeStamp class, but rather the Base interface. But, you can see how it all might work.

```ts
@ModelOptions({
  schemaOptions: {
    timestamps: { updatedAt: 'modifiedAt' },
    toObject: { virtuals: true }
  }
})
@Plugin(mongooseAutopopulate as any)
export class BaseEntity implements Base {
  _id!: Types.ObjectId

  id!: string

  @Prop({ ref: () => UserEntity, autopopulate: { maxDepth: 1 } })
  public createdBy?: Ref<UserEntity>

  @Prop({ ref: () => UserEntity, autopopulate: { maxDepth: 1 } })
  public modifiedBy?: Ref<UserEntity>

  @Prop()
  public modifiedAt!: Date

  @Prop()
  public createdAt!: Date
}
```

**Comment by hasezoey** [comment, marked as answer]
> should be automatically created

they are

> However, they are marked optional in types

they are marked this way, otherwise (with `!`) then typescript would require you to add there variables in an `.create` call (maybe just in `strict` tsconfig mode)
